### PR TITLE
Fixes #210, #211, #212

### LINKED ISSUE #210: parser: parseExpr4 runs in infinite recursion
**Describe the bug**
Fuzzing fails.

**To Reproduce**
Parse any of the attached crasher files.

Crash files are attached
[Archive.zip](https://github.com/tomarrell/lbadd/files/4947340/Archive.zip)
[Archive 2.zip](https://github.com/tomarrell/lbadd/files/4947346/Archive.2.zip)
[Archive 3.zip](https://github.com/tomarrell/lbadd/files/4947348/Archive.3.zip)
[Archive 4.zip](https://github.com/tomarrell/lbadd/files/4947349/Archive.4.zip)

**Expected behavior**
No crash

### PR DESCRIPTION
**Pull request by SUMUKHA-PK**
This PR discovers that some statements are not valid utf8 strings and adds support for gracefully rejecting those inputs.
Changes in all files are done to accommodate the same.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,6 +3,8 @@ module github.com/tomarrell/lbadd
 go 1.13
 
 require (
+	github.com/davecgh/go-spew v1.1.1
+	github.com/dvyukov/go-fuzz v0.0.0-20200318091601-be3528f3a813 // indirect
```
Comment (TimSatke) [inline go.mod:7]: run `go mod tidy` before committing please

File: go.mod342847743.tmp [hunk 2]
```diff
@@ -0,0 +1,22 @@
+module github.com/tomarrell/lbadd
```
Comment (TimSatke) [inline go.mod342847743.tmp:1]: what is this file?

File: internal/compiler/golden_test.go [hunk 3]
```diff
@@ -36,7 +37,8 @@ func runGolden(t *testing.T, input string) {
 	require := require.New(t)
 
 	c := &simpleCompiler{}
-	p := parser.New(input)
+	p, err := parser.New(input)
+	assert.Nil(t, err)
```
Comment (TimSatke) [inline internal/compiler/golden_test.go:41]:
```suggestion
	require.NoError(err)
```

File: internal/compiler/simple_compiler_test.go [hunk 4]
```diff
@@ -729,7 +729,9 @@ func _TestCompile(tt testcase) func(t *testing.T) {
 		assert := assert.New(t)
 
 		c := &simpleCompiler{}
-		p := parser.New(tt.input)
+		p, err := parser.New(tt.input)
+		assert.Nil(err)
```
Comment (TimSatke) [inline internal/compiler/simple_compiler_test.go:733]:
```suggestion
		assert.NoError(err)
```

File: internal/parser/scanner/rule_based_scanner_test.go [hunk 5]
```diff
@@ -169,26 +169,26 @@ func TestRuleBasedScanner(t *testing.T) {
 		},
 		{
 			"numeric literals",
-			"7 7.5 8.9.8 8.0 0.4 10 10000 18907.890 1890976.09.977",
+			"7 7.5 8.9 .8 8.0 0.4 10 10000 18907.890 1890976.09 .977",
```
Comment (TimSatke) [inline internal/parser/scanner/rule_based_scanner_test.go:172]: why did you change these tests?

File: internal/parser/scanner/rule_based_scanner_test.go [hunk 6]
```diff
@@ -309,7 +309,8 @@ func _TestRuleBasedScannerWithRuleset(input string, ruleset ruleset.Ruleset, wan
 		var got []token.Token
 
 		// create the scanner to be tested
-		sc := NewRuleBased([]rune(input), ruleset)
+		sc, err := NewRuleBased(input, ruleset)
+		assert.Nil(err)
```
Comment (TimSatke) [inline internal/parser/scanner/rule_based_scanner_test.go:313]:
```suggestion
		assert.NoError(err)
```

File: internal/parser/scanner/rule_based_scanner_test.go [hunk 7]
```diff
@@ -336,11 +337,12 @@ func _TestRuleBasedScannerWithRuleset(input string, ruleset ruleset.Ruleset, wan
...
 }
 
-func TestRuleBasedSannerStatementEndingInWhitespace(t *testing.T) {
+func TestRuleBasedScannerStatementEndingInWhitespace(t *testing.T) {
 	assert := assert.New(t)
 
 	stmt := "SELECT "
-	sc := NewRuleBased([]rune(stmt), ruleset.Default)
+	sc, err := NewRuleBased(stmt, ruleset.Default)
+	assert.Nil(err)
```
Comment (TimSatke) [inline internal/parser/scanner/rule_based_scanner_test.go:345]:
```suggestion
	assert.NoError(err)
```

File: internal/parser/scanner/scanner.go [hunk 8]
```diff
@@ -11,7 +11,8 @@ func (s Error) Error() string { return string(s) }
 
 // Constant errors
 const (
-	ErrUnexpectedToken = Error("unexpected token")
+	ErrUnexpectedToken   = Error("unexpected token")
+	ErrInvalidUTF8String = Error("the input is not a valid utf8 encoded string")
```
Comment (TimSatke) [inline internal/parser/scanner/scanner.go:15]:
```suggestion
	ErrInvalidUTF8String = Error("input is not a valid utf8 encoded string")
```

File: internal/parser/simple_parser_rules.go [hunk 9]
```diff
@@ -3642,6 +3660,11 @@ func (p *simpleParser) parseQualifiedTableName(r reporter) (stmt *ast.QualifiedT
 			r.unexpectedToken(token.KeywordIndexed)
 		}
 	}
+
+	if reflect.DeepEqual(stmt, &ast.QualifiedTableName{}) {
+		return nil
+	}
```
Comment (TimSatke) [inline internal/parser/simple_parser_rules.go:3666]: Don't use reflect here, please.
```suggestion
	if (*stmt == ast.QualifiedTableName{}) {
		return nil
	}
```

File: internal/test/base_test.go [hunk 10]
```diff
@@ -72,7 +72,9 @@ func runAndCompare(t *testing.T, tt Test) {
 	totalStart := time.Now()
 	parseStart := time.Now()
 
-	p := parser.New(tt.Statement)
+	p, err := parser.New(tt.Statement)
+	assert.Nil(err)
```
Comment (TimSatke) [inline internal/test/base_test.go:76]:
```suggestion
	assert.NoError(err)
```

File: internal/parser/parser_test.go [hunk 11]
```diff
@@ -15,9811 +15,9811 @@ func TestSingleStatementParse(t *testing.T) {
...
-					WithClause: &ast.WithClause{
-						With: token.New(1, 1, 0, 4, token.KeywordWith, "WITH"),
-						RecursiveCte: []*ast.RecursiveCte{
-							{
-								CteTableName: &ast.CteTableName{
-									TableName: token.New(1, 6, 5, 7, token.Literal, "myTable"),
-								},
-								As:        token.New(1, 14, 13, 2, token.KeywordAs, "AS"),
-								LeftParen: token.New(1, 17, 16, 1, token.Delimiter, "("),
+						{
```
Comment (TimSatke) [inline internal/parser/parser_test.go]: I allowed myself to fix that indentation for you

File: internal/parser/parser_test.go [hunk 12]
```diff
@@ -9818,8 +9818,8 @@ func TestSingleStatementParse(t *testing.T) {
 		t.Run(input.Name, func(t *testing.T) {
 			assert := assert.New(t)
 
-			p := New(input.Query)
-
+			p, err := New(input.Query)
+			assert.Nil(err)
```
Comment (TimSatke) [inline internal/parser/parser_test.go:9822]:
```suggestion
			assert.NoError(err)
```

File: internal/parser/scanner/rule_based_scanner_test.go (hunk 5 above)
Comment (SUMUKHA-PK) [reply internal/parser/scanner/rule_based_scanner_test.go:172]: Similar ones were existing already.

Comment (SUMUKHA-PK) [reply internal/parser/scanner/rule_based_scanner_test.go:172]: Reverted. Doesnt make much of a difference, those points are covered.

File: internal/parser/simple_parser_rules.go (hunk 9 above)
Comment (SUMUKHA-PK) [reply internal/parser/simple_parser_rules.go:3666]: Ah I didnt know how to compare them that way. The brackets were the issue. Fixed.

**Comment by SUMUKHA-PK** [comment]
Done.
